cleanlab_pinecone_rag_workflow/Cleanlab_Pinecone_RAG.ipynb

```diff
@@ -0,0 +1,2673 @@
+    "            batch_metadata = valid_metadata[i:i+batch_size]\n",
+    "            batch_embeddings = doc_embeddings[i:i+batch_size]\n",
+    "            \n",
+    "            vectors = [\n",
+    "                (generated_ids[i+j], embedding.tolist(), {**metadata, 'text': doc[:1000]})\n",
+    "                for j, (doc, embedding, metadata) in enumerate(zip(batch_docs, batch_embeddings, batch_metadata))\n",
+    "            ]\n",
+    "\n",
+    "            try:\n",
+    "                self.index.upsert(vectors=vectors)\n",
+    "                print(f\"Successfully indexed batch of {len(vectors)} chunks.\")\n",
+    "            except Exception as e:\n",
+    "                print(f\"Error during upsert: {e}\")\n",
+    "\n",
+    "        print(\"Document indexing completed.\")\n",
+    "        return len(valid_docs)\n",
+    "\n",
+    "    def search(self, query: str, top_k: int = 5, filter_query: Optional[Dict] = None) -> List[Tuple[str, Dict]]:\n",
+    "        \"\"\"\n",
+    "        Search for the top_k most relevant documents based on the input query and optional filter.\n",
+    "        \n",
+    "        Args:\n",
+    "            query (str): The search query text.\n",
+    "            top_k (int): The number of top relevant documents to return.\n",
+    "            filter_query (Optional[Dict]): Optional filter query to apply during search.\n",
+    "        \n",
+    "        Returns:\n",
+    "            List[Tuple[str, Dict]]: List of top_k relevant document texts and their metadata.\n",
+    "                                    Each tuple contains (document_text, metadata_dict).\n",
+    "        \"\"\"\n",
+    "        query_embedding = self.model.encode(query)\n",
+    "        \n",
+    "        try:\n",
+    "            results = self.index.query(\n",
+    "                vector=query_embedding.tolist(),\n",
+    "                top_k=top_k,\n",
+    "                filter=filter_query,\n",
+    "                include_metadata=True\n",
+    "            )\n",
+    "            \n",
+    "            return [(match.metadata['text'], {k: v for k, v in match.metadata.items() if k != 'text'}) for match in results.matches]\n",
+    "        except Exception as e:\n",
+    "            print(f\"Error during search: {e}\")\n",
+    "            return []\n",
+    "\n",
+    "    def delete_index(self) -> None:\n",
+    "        \"\"\"\n",
+    "        Delete the Pinecone index.\n",
+    "        \n",
+    "        Raises:\n",
+    "            Exception: If there's an error during the deletion process.\n",
+    "        \"\"\"\n",
+    "        try:\n",
+    "            self.pc.delete_index(self.index_name)\n",
+    "            print(f\"Index '{self.index_name}' has been deleted.\")\n",
+    "        except Exception as e:\n",
+    "            print(f\"Error deleting index: {e}\")\n",
+    "\n",
+    "    def extract_chunks_from_index(self) -> pd.DataFrame:\n",
```

Comment:
Nit this should be named more clearly...I think this is "Extract the top 10,000 chunks from the vector that are closest to the origin"? 

Typically a function called something like `extract_chunks` is going to take in a string as input and do a similarity search, returning `top_k` related chunks... so I feel this is very confusing to RAG practioners